include/openthread/thread.h

```diff
@@ -264,6 +264,22 @@ otError otThreadDiscover(otInstance *             aInstance,
  */
 bool otThreadIsDiscoverInProgress(otInstance *aInstance);
 
+/**
+ * Set the Thread Joiner Advertisement when discovering Thread network.
+ *
+ * @param[in]  aInstance A pointer to an OpenThread instance.
+ * @param[in]  aOui      The Vendor OUI value which will be included in the Joiner Advertisement.
+ * @param[in]  aAdvData  The AdvData which will be included in the Joiner Advertisement.
+ *
+ * @retval OT_ERROR_NONE         Successfully set Joiner Advertisement.
+ * @retval OT_ERROR_INVALID_ARGS Invalid AdvData.
+ *
+ */
+otError otThreadSetJoinerAdvertisement(otInstance *   aInstance,
+                                       uint32_t       aOui,
+                                       const uint8_t *aAdvData,
+                                       uint8_t        aAdvDataLength);
+
```

Comment:
Looks like th current code assumes that the caller is provding the buffer for `aAdvData` (i.e. the content is not copied an just the pointer is saved).

There are pros/cons for this model (mainly that it makes it more complex for caller/user of OT), but I think it is good and flexible model for the initial implementation.


_Assuming we choose to go for this model (i.e., caller need to provides the buffer/object)_, I suggest we change it as follows:
- We define a new type `otJoinerAdvData` 
```
/**
 * This structure represents a Joiner Advertisement Data.
 *
 */
typedef struct otJoinerAdvData
{
    uint32_t                      mOui;        ///< Vendor OUI
    const uint8_t *               mData;       ///< Pointer to buffer containing the Adv Data
    uint8_t                       mDataLength; ///< The length of data (in bytes)
    const struct otJoinerAdvData *mNext;       ///< For internal OT core use.
} otJoinerAdvData;
```
- We define `otThreadAddJoinerAdvData(otInstance *, const otJoinAdvData *)` and `otThreadRemoveAdvData()` to allow user to add and remove entries (e.g., associated with different OUIs)
- We require the `otJoinerAdvData` instance that the caller provdies to remain valid.
- Within OT core just put the given objects into a linked-list.

Thoughts?

Reply:
Thanks for the suggestion Abtin. I've changed the code to copy the content rather than just save the pointer as the latter approach could be problematic if the user fails to keep the data valid until the TLV is sent. Please take a look.

Reply:
SGTM